server/src/org/apache/cloudstack/network/lb/CertServiceImpl.java

```diff
@@ -341,179 +350,172 @@ public SslCertResponse createCertResponse(SslCertVO cert, List<LoadBalancerCertM
             throw new InvalidParameterValueException("Invalid Certificate format. Expected X509 certificate. Failed due to " + e.getMessage());
         } finally {
             IOUtils.closeQuietly(certPem);
         }
     }
 
-    public List<Certificate> parseChain(String chain) throws IOException {
+    private Certificate readCertificateFromPemObject(PemObject pemObject) throws CertificateException {
+        final ByteArrayInputStream bais = new ByteArrayInputStream(pemObject.getContent());
+        final CertificateFactory certificateFactory = CertificateFactory.getInstance("X509");
 
-        List<Certificate> certs = new ArrayList<Certificate>();
-        PEMReader reader = new PEMReader(new StringReader(chain));
+        return certificateFactory.generateCertificate(bais);
+    }
 
-        Certificate crt = null;
 
-        while ((crt = (Certificate)reader.readObject()) != null) {
-            if (crt instanceof X509Certificate) {
-                certs.add(crt);
-            }
-        }
-        if (certs.size() == 0)
-            throw new IllegalArgumentException("Unable to decode certificate chain");
+    public List<Certificate> parseChain(String chain) throws IOException, CertificateException {
 
-        return certs;
+        return CertificateHelper.parseChain(chain);
     }
 
     String generateFingerPrint(Certificate cert) {
 
         final char[] HEX = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
 
-        StringBuilder buffer = new StringBuilder(60);
+        final StringBuilder buffer = new StringBuilder(60);
         try {
 
-            MessageDigest md = MessageDigest.getInstance("SHA-1");
-            byte[] data = md.digest(cert.getEncoded());
+            final MessageDigest md = MessageDigest.getInstance("SHA-1");
+            final byte[] data = md.digest(cert.getEncoded());
 
-            for (int i = 0; i < data.length; i++) {
+            for (final byte element : data) {
                 if (buffer.length() > 0) {
                     buffer.append(":");
                 }
 
-                buffer.append(HEX[(0xF0 & data[i]) >>> 4]);
-                buffer.append(HEX[0x0F & data[i]]);
+                buffer.append(HEX[(0xF0 & element) >>> 4]);
+                buffer.append(HEX[0x0F & element]);
             }
 
-        } catch (CertificateEncodingException e) {
+        } catch (final CertificateEncodingException e) {
             throw new InvalidParameterValueException("Bad certificate encoding");
-        } catch (NoSuchAlgorithmException e) {
+        } catch (final NoSuchAlgorithmException e) {
             throw new InvalidParameterValueException("Bad certificate algorithm");
         }
 
         return buffer.toString();
     }
 
-    public static class KeyPassword implements PasswordFinder {
+    /**
+     *
+     * @deprecated this is only for bcprov-jdk16
+     *
+     */
+    @Deprecated
+    public static class KeyPassword {
```

Comment:
If it's not in use, why deprecate it?  The reason for deprecation is to allow existing code to evolve at lower risk.  If it's not used, there seems to be no good reason to keep it and the downside risk that someone will ignore the deprecation warning and use it.